topology/src/main/java/io/camunda/zeebe/topology/changes/TopologyChangeCoordinatorImpl.java

```diff
@@ -0,0 +1,203 @@
+import io.camunda.zeebe.topology.changes.TopologyChangeAppliers.OperationApplier;
+import io.camunda.zeebe.topology.state.ClusterTopology;
+import io.camunda.zeebe.topology.state.TopologyChangeOperation;
+import java.util.ConcurrentModificationException;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TopologyChangeCoordinatorImpl implements TopologyChangeCoordinator {
+  private static final Logger LOG = LoggerFactory.getLogger(TopologyChangeCoordinatorImpl.class);
+
+  private final ClusterTopologyManager clusterTopologyManager;
+
+  private final ConcurrencyControl executor;
+
+  public TopologyChangeCoordinatorImpl(
+      final ClusterTopologyManager clusterTopologyManager, final ConcurrencyControl executor) {
+    this.clusterTopologyManager = clusterTopologyManager;
+    this.executor = executor;
+  }
+
+  @Override
+  public ActorFuture<ClusterTopology> applyOperations(
+      final List<TopologyChangeOperation> operations) {
+    final ActorFuture<ClusterTopology> future = executor.createFuture();
+
+    clusterTopologyManager
+        .getClusterTopology()
+        .onComplete(
+            (currentClusterTopology, errorOnGettingTopology) -> {
+              if (errorOnGettingTopology != null) {
+                future.completeExceptionally(errorOnGettingTopology);
+                return;
+              }
+
+              final ActorFuture<ClusterTopology> validation =
+                  validateTopologyChangeRequest(currentClusterTopology, operations);
+
+              validation.onComplete(
+                  (simulatedFinalTopology, validationError) -> {
+                    if (validationError != null) {
+                      future.completeExceptionally(validationError);
+                      return;
+                    }
+
+                    // if the validation was successful, apply the changes
+                    applyTopologyChange(
+                        operations, currentClusterTopology, simulatedFinalTopology, future);
+                  });
+            });
+    return future;
+  }
+
+  @Override
+  public ActorFuture<Boolean> hasCompletedChanges(final long version) {
+    final ActorFuture<Boolean> future = executor.createFuture();
+    clusterTopologyManager
+        .getClusterTopology()
+        .onComplete(
+            (currentClusterTopology, error) -> {
+              if (error != null) {
+                future.completeExceptionally(error);
+                return;
+              }
+
+              if (currentClusterTopology.version() == version) {
+                future.complete(!currentClusterTopology.hasPendingChanges());
+              } else {
+                future.completeExceptionally(
+                    new UnknownStatus(
+                        String.format(
+                            "The topology has changed since the version %d. The current version is %d. The topology change would have been already completed.",
```

Comment:
:upside_down_face: Technically not correct, if `version < currentClusterTopology.version()` we could show a different error message